Add support for Web Extension cookies API.

xeenon opened the pull request:
#### 2ffa7c91980ec5d1b448b38896c59dac4b77819c
<pre>
Add support for Web Extension cookies API.
<a href="https://webkit.org/b/261771">https://webkit.org/b/261771</a>
<a href="https://rdar.apple.com/problem/115743156">rdar://problem/115743156</a>

Reviewed by Brian Weinstein.

Finish up the support for cookies, which includes support for multiple cookies stores.

This required new _WKWebExtensionControllerConfiguration property to define what the
default WKWebsiteDataStore is, and the other stores are tracked by what tabs are open.

Added a bunch of new tests that cover all the cookies APIs.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(class_template_headers):
(headers_for_type):
* Source/WebKit/Shared/Extensions/WebExtensionCookieParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionCookieParameters.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::validateSingleObject):
(WebKit::validateDictionary):
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::cookies):
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(-[_WKWebExtensionControllerConfiguration encodeWithCoder:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
(-[_WKWebExtensionControllerConfiguration defaultWebsiteDataStore]):
(-[_WKWebExtensionControllerConfiguration setDefaultWebsiteDataStore:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICookiesCocoa.mm: Added.
(WebKit::domainsMatch):
(WebKit::toURL):
(WebKit::WebExtensionContext::fetchCookies):
(WebKit::WebExtensionContext::cookiesGet):
(WebKit::WebExtensionContext::cookiesGetAll):
(WebKit::WebExtensionContext::cookiesSet):
(WebKit::WebExtensionContext::cookiesRemove):
(WebKit::WebExtensionContext::cookiesGetAllCookieStores):
(WebKit::WebExtensionContext::fireCookiesChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::websiteDataStore const):
(WebKit::WebExtensionContext::cookiesDidChange):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::addPage):
(WebKit::WebExtensionController::removePage):
(WebKit::WebExtensionController::websiteDataStore const):
(WebKit::WebExtensionController::addWebsiteDataStore):
(WebKit::WebExtensionController::removeWebsiteDataStore):
(WebKit::WebExtensionController::cookiesDidChange):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::copy const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::allPages const):
(WebKit::WebExtensionController::allWebsiteDataStores const):
(WebKit::WebExtensionController::allUserContentControllers const):
(WebKit::WebExtensionController::allNonPrivateUserContentControllers const):
(WebKit::WebExtensionController::allPrivateUserContentControllers const):
(WebKit::WebExtensionController::allProcessPools const):
(WebKit::WebExtensionController::HTTPCookieStoreObserver::HTTPCookieStoreObserver):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp:
(WebKit::WebExtensionControllerConfiguration::operator== const):
(WebKit::WebExtensionControllerConfiguration::defaultWebsiteDataStore const):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::setDefaultWebsiteDataStore):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::operator== const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICookiesCocoa.mm:
(WebKit::toImpl):
(WebKit::toWebAPI):
(WebKit::WebExtensionAPICookies::parseCookieDetails):
(WebKit::WebExtensionAPICookies::get):
(WebKit::normalizeDomain):
(WebKit::WebExtensionAPICookies::getAll):
(WebKit::WebExtensionAPICookies::set):
(WebKit::WebExtensionAPICookies::remove):
(WebKit::WebExtensionAPICookies::getAllCookieStores):
(WebKit::WebExtensionContextProxy::dispatchCookiesChangedEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPICookies.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICookies.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionControllerConfiguration.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/273160@main">https://commits.webkit.org/273160@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d144e99355ac1bf1b2279527e551bf827e33b35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34300 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13109 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36293 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36980 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31053 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35382 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30066 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34812 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11093 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30591 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9684 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/34599 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9781 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38267 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31156 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30958 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35860 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9908 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7780 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33786 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11703 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10465 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4426 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10743 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/Shared/Extensions/WebExtensionCookieParameters.h:
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
2024

kiaraarose commented on Source/WebKit/Shared/Extensions/WebExtensionCookieParameters.serialization.in:
```diff
@@ -0,0 +1,38 @@
+# Copyright (C) 2023 Apple Inc. All rights reserved.
```
2024

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICookiesCocoa.mm:
```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
2024

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2346,6 +2346,19 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    return result.get();
+}
+
+void WebExtensionContext::cookiesDidChange(API::HTTPCookieStore& cookieStore)
```
should we be doing something with `cookieStore` here?

xeenon replied:
We can't. I'll put a FIXME here too.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
```diff
@@ -48,6 +48,13 @@ messages -> WebExtensionContext {
     CommandsGetAll() -> (Vector<WebKit::WebExtensionCommandParameters> commands);
 
+    // Cookies
+    CookiesGet(std::optional<PAL::SessionID> sessionID, String name, URL url) -> (std::optional<WebKit::WebExtensionCookieParameters> cookie, std::optional<String> error);
```
`WebKit::WebExtensionContext:ErrorString` instead of  ` std::optional<String> error`? more wordy but it might be better to use here

xeenon replied:
I'll try.

xeenon committed:
Add support for Web Extension cookies API.
https://webkit.org/b/261771
rdar://problem/115743156

Reviewed by Brian Weinstein.

Finish up the support for cookies, which includes support for multiple cookies stores.

This required new _WKWebExtensionControllerConfiguration property to define what the
default WKWebsiteDataStore is, and the other stores are tracked by what tabs are open.

Added a bunch of new tests that cover all the cookies APIs.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(class_template_headers):
(headers_for_type):
* Source/WebKit/Shared/Extensions/WebExtensionCookieParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionCookieParameters.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::validateSingleObject):
(WebKit::validateDictionary):
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::cookies):
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(-[_WKWebExtensionControllerConfiguration encodeWithCoder:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
(-[_WKWebExtensionControllerConfiguration defaultWebsiteDataStore]):
(-[_WKWebExtensionControllerConfiguration setDefaultWebsiteDataStore:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICookiesCocoa.mm: Added.
(WebKit::domainsMatch):
(WebKit::toURL):
(WebKit::WebExtensionContext::fetchCookies):
(WebKit::WebExtensionContext::cookiesGet):
(WebKit::WebExtensionContext::cookiesGetAll):
(WebKit::WebExtensionContext::cookiesSet):
(WebKit::WebExtensionContext::cookiesRemove):
(WebKit::WebExtensionContext::cookiesGetAllCookieStores):
(WebKit::WebExtensionContext::fireCookiesChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::websiteDataStore const):
(WebKit::WebExtensionContext::cookiesDidChange):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::addPage):
(WebKit::WebExtensionController::removePage):
(WebKit::WebExtensionController::websiteDataStore const):
(WebKit::WebExtensionController::addWebsiteDataStore):
(WebKit::WebExtensionController::removeWebsiteDataStore):
(WebKit::WebExtensionController::cookiesDidChange):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::copy const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::allPages const):
(WebKit::WebExtensionController::allWebsiteDataStores const):
(WebKit::WebExtensionController::allUserContentControllers const):
(WebKit::WebExtensionController::allNonPrivateUserContentControllers const):
(WebKit::WebExtensionController::allPrivateUserContentControllers const):
(WebKit::WebExtensionController::allProcessPools const):
(WebKit::WebExtensionController::HTTPCookieStoreObserver::HTTPCookieStoreObserver):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp:
(WebKit::WebExtensionControllerConfiguration::operator== const):
(WebKit::WebExtensionControllerConfiguration::defaultWebsiteDataStore const):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::setDefaultWebsiteDataStore):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
(WebKit::WebsiteDataStore::operator== const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICookiesCocoa.mm:
(WebKit::toImpl):
(WebKit::toWebAPI):
(WebKit::WebExtensionAPICookies::parseCookieDetails):
(WebKit::WebExtensionAPICookies::get):
(WebKit::normalizeDomain):
(WebKit::WebExtensionAPICookies::getAll):
(WebKit::WebExtensionAPICookies::set):
(WebKit::WebExtensionAPICookies::remove):
(WebKit::WebExtensionAPICookies::getAllCookieStores):
(WebKit::WebExtensionContextProxy::dispatchCookiesChangedEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPICookies.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICookies.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionControllerConfiguration.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/273160@main
